src/app/clusters/scenes-server/SceneTableImpl.cpp

```diff
@@ -197,35 +198,37 @@ struct SceneTableData : public SceneTableEntry, PersistentData<kPersistentSceneB
 static constexpr size_t kPersistentFabricBufferMax = 128;
 
 /**
- * @brief Linked list of all scenes in a fabric, stored in persistent memory
+ * @brief Class that holds a map to all scenes in a fabric for a specific endpoint
  *
  * FabricSceneData is an access to a linked list of scenes
  */
 struct FabricSceneData : public PersistentData<kPersistentFabricBufferMax>
 {
+    EndpointId endpoint_id;
     FabricIndex fabric_index;
     uint8_t scene_count = 0;
-    uint8_t max_scenes_per_fabric;
-    uint8_t max_scenes_global;
+    uint16_t max_scenes_per_fabric;
+    uint16_t max_scenes_per_endpoint;
     SceneStorageId scene_map[kMaxScenesPerFabric];
 
-    FabricSceneData(FabricIndex fabric = kUndefinedFabricIndex, uint8_t maxScenesPerFabric = kMaxScenesPerFabric,
-                    uint8_t maxScenesGlobal = kMaxScenesGlobal) :
-        fabric_index(fabric),
-        max_scenes_per_fabric(maxScenesPerFabric), max_scenes_global(maxScenesGlobal)
+    FabricSceneData(EndpointId endpoint = kInvalidEndpointId, FabricIndex fabric = kUndefinedFabricIndex,
+                    uint16_t maxScenesPerFabric = kMaxScenesPerFabric, uint16_t maxScenesPerEndpoint = kMaxScenesPerEndpoint) :
+        endpoint_id(endpoint),
+        fabric_index(fabric), max_scenes_per_fabric(maxScenesPerFabric), max_scenes_per_endpoint(maxScenesPerEndpoint)
     {}
 
     CHIP_ERROR UpdateKey(StorageKeyName & key) override
     {
         VerifyOrReturnError(kUndefinedFabricIndex != fabric_index, CHIP_ERROR_INVALID_FABRIC_INDEX);
-        key = DefaultStorageKeyAllocator::FabricSceneDataKey(fabric_index);
+        VerifyOrReturnError(kInvalidEndpointId != endpoint_id, CHIP_ERROR_INVALID_ARGUMENT);
+        key = DefaultStorageKeyAllocator::FabricSceneDataKey(fabric_index, endpoint_id);
         return CHIP_NO_ERROR;
     }
 
     void Clear() override
     {
         scene_count = 0;
-        for (uint8_t i = 0; i < max_scenes_per_fabric; i++)
+        for (uint16_t i = 0; i < max_scenes_per_fabric; i++)
```

Comment:
For future reference: all these loops would be a lot safer with a `using SceneIndexType = uint16_t` somewhere and then not having to keep a bunch of types in sync.  Again, followup is just fine if we want to make this change.

Reply:
SceneIndex is now a uint16_t.

Reply:
Create [issue](https://github.com/orgs/project-chip/projects/19?pane=issue&itemId=29272989) for followup.